torch/ao/quantization/experimental/linear.py

```diff
@@ -0,0 +1,168 @@
+
+                    for idx in range(len(weight_val)):
+                        ele = int(weight_val[idx])
+
+                        x = len(weight_val) - 1 - idx
+
+                        if ele:
+                            curr_result = r << x
+                        else:
+                            curr_result = 0
+                        result[i][j] += curr_result
+
+        return result
+
+    def matmul(self, decomposed_weight, activation):
+        r"""
+        Call bitshift_mul function to perform matrix multiplication between
+        decomposed_weight and activation.
+        Args:
+            decomposed_weight (Tensor): APoT quantized weight decomposed into binary
+            activation (Tensor): uniformly quantized activation
+        """
+        return self.bitshift_mul(decomposed_weight, activation)
+
+    def linear_APoT_fn(self, activation: torch.Tensor) -> torch.FloatTensor:
+        r"""
+        Multiply APoT quantized weight and uniformly quantized activation (dtype: quint8)
+        with bitshifting instead of matrix multiplication.
+        Result has dtype torch.float32
+        Args:
+            activation (Tensor): uniformly quantized activation tensor
+        """
+        assert activation.dim() == 2
+
+        weight_rows = self.weight_transposed.size()[0]
+        weight_cols = self.weight_transposed.size()[1]
+
+        decomposed_weight = np.empty(shape=(weight_rows, weight_cols), dtype=object)
+        for row in range(weight_rows):
+            for col in range(weight_cols):
+                decomposed_weight[row][col] = self.decompose_APoT(bin(self.weight_transposed[row][col]))
+
+        rows1 = self.weight_transposed.size(dim=0)
+        cols1 = self.weight_transposed.size(dim=1)
+
+        rows2 = activation.size(dim=0)
+        cols2 = activation.size(dim=1)
+
+        result = self.matmul(decomposed_weight, activation).type(torch.FloatTensor)
+
+        return result
+
+    def forward(self, activation: torch.Tensor) -> torch.FloatTensor:
+        r"""
+        Call linear_APoT_fn to multiply activation with an APoT quantized weight.
+        Args:
+            activation (Tensor): uniformly quantized activation tensor
+        """
+        return self.linear_APoT_fn(activation)
+
+    @classmethod
+    def from_reference(cls,  # type: ignore[override]
+                       ref_qlinear,
+                       alpha: torch.Tensor,
+                       gamma: torch.Tensor,
+                       quantization_levels: torch.Tensor,
+                       level_indices: torch.Tensor):
+        r"""Create a (fbgemm/qnnpack) quantized module from a reference quantized module
+
+        Args:
+            ref_qlinear (Module): a reference quantized linear module, either
+                                  produced by torch.ao.quantization.experimental
+                                  utilities or provided by the user
+        """
+        qlinear = ref_qlinear
```

Comment:
I think we'll need to create a `LinearAPoT` instance here, for example: https://github.com/pytorch/pytorch/blob/master/torch/nn/quantized/modules/linear.py#L290

Reply:
also this path is not yet tested, maybe we can leave this unimplemented for now

Reply:
sounds good, will leave unimplemented